Collect constant from rust-bitcoin projects into one library

sgeisler opened the pull request:
extract constants from:

 * rust-bech32-bitcoin
 * rust-bitcoin

and combine them in one crate, so we can avoid redundant implementations of the same constants.

sgeisler committed:
merge constant from rust-bitcoin projects

extract constants from:
 * rust-bech32-bitcoin
 * rust-bitcoin
and combine them in one crate to avoid redundant implementations.

sgeisler commented on src/lib.rs:
```diff
@@ -0,0 +1,229 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        let name = match *self {
+            Network::Bitcoin => "bitcoin",
+            Network::Testnet => "testnet", // only 'testnet' for compatibility reasons
```
I didn't change the representation of the bitcoin testnet for compatibility reasons, but `bitcoin-testnet` would be the much better encoding if possible.

apoelstra replied:
I think it's fine to change.

sgeisler commented on Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+serde = { version = "0.6", optional = true }
+
+[dev-dependencies]
+serde_json = "0.6"
```
The version of `serde_json` is rather arbitrary: I just tried till something worked with 1.14.0 so I could test the serde `impl`s

apoelstra replied:
There is a `serde_test` crate that you can use for unit tests. I think for "real" serialization tests it'd be better to write them as a fuzzer - then you can use serde 1.0 because (I think) we have no compiler requirements for fuzztesting.

sgeisler commented on Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "bitcoin-constants"
+version = "0.1.0"
+authors = ["The rust-bitcoin developers"]
```
Since most of the code wasn't from me/or another single person I used a more generic author description.

sgeisler committed:
Define trait for network constant enums

clarkmoody commented on src/lib.rs:
```diff
@@ -0,0 +1,301 @@
+// Copyright (c) 2017 Clark Moody
```
This has been developed quite a bit more than my original contribution. Could add or change to "The rust-bitcoin developers"

sgeisler replied:
Oh, I just overlooked that one.

sgeisler committed:
Add chain parameters for bitcoin

(main, test, reg)

sgeisler committed:
Add genesis block hashes for bitcoin

sgeisler committed:
add address prefix constants

sgeisler committed:
Add safe subset of networks with guaranteed complete implementation

Some constants aren't implemented for all networks. This PR defines
a safe subset (namely bitcoin, bitcoin testnet, bitcoin regtest)
that will no panic because of unimplemented constants.

sgeisler committed:
add xpub/xpriv prefixes

sgeisler committed:
add wif prefixes

sgeisler commented on src/lib.rs:
```diff
@@ -0,0 +1,652 @@
+///
+/// If you feel the urge to do matching over an enum implementing `NetworkConstants` please
+/// consider opening a PR instead if your problem/solution can be generalized.
+pub trait NetworkConstants : Sized {
```
After adding all constants I could find in `rust-bitcoin` this trait has become rather large. One possibility would be to split it into formatting (hrp, prefixes, ...), consensus (chain params, genesis block) and hodgepodge (name, network magic bytes).

My reasoning behind this would be that nearly no function which takes a network as parameter will need all of it's constants. But if you pass a trait reference the v-table will be quite large due to the many functions.

stevenroose commented on Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+name = "bitcoin-constants"
+version = "0.1.0"
+authors = ["The rust-bitcoin developers"]
+repository = "https://github.com/rust-bitcoin/bitcoin-constants"
```
If we dropped `rust-` because it's already in the org name, shouldn't this repo be named `constants`? :D

sgeisler replied:
I guess that would be a quite deceiving name on crates.io :smile:

stevenroose replied:
Not talking about the crate name though, just the repo name.

sgeisler replied:
Ok, that would make sense if we could implement this name change for all crates. Otherwise it will only decrease consistency. But I can't change the name anyway.

stevenroose replied:
Yeah I've been asking @apoelstra :) Only this one and hashes are the ones deviating..

apoelstra replied:
Yeah, sure, I'll rename it.

stevenroose commented on Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+version = "0.1.0"
+authors = ["The rust-bitcoin developers"]
+repository = "https://github.com/rust-bitcoin/bitcoin-constants"
+description = "Contains network constants for bitcoin and similiar crypto currencies"
```
We're really gonna do other cryptocurrencies?

sgeisler replied:
Well, `rust-bech32-bitcoin` does include constants for vertcoin and litecoin iirc, that's how they ended up in this PR. If @clarkmoody would be willing to drop them I'd remove them from this PR but would still like to keep the extensible structure (enum implementing trait).

clarkmoody replied:
I'm fine with dropping support for the other coins, until we get serious requests from other projects.

stevenroose commented on Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+readme = "README.md"
+keywords = ["bitcoin", "constants", "blockchain"]
+categories = []
+license = "MIT"
```
rust-bitcoin is CC0, probably better stick to that, no?

stevenroose replied:
You're also missing the license file.

sgeisler replied:
Oh, yeah, I don't care as long as it's not GPL-like, CC0 seems to be even more laissez-faire than MIT, so that's OK with me. Maybe we should ask @clarkmoody since he uses MIT for `rust-bech32-bitcoin` and I used parts of it (even though they changed so much by now ...).

clarkmoody replied:
Just defaulted to MIT, since that was Bitcoin's license. But this code bears nearly no resemblance to my original.

apoelstra replied:
@clarkmoody would you be willing to relicense under CC0 for better compatibility with the rust-bitcoin license? (Maybe you can't do this .. I'm not sure how many contributors you have already.)

clarkmoody replied:
I'll open a PR to re-license `rust-bech32-bitcoin` and see if existing contributors are alright

stevenroose reviewed (changes requested):
I really don't see much value is trying to contain non-Bitcoin data in rust-bitcoin. The `ChainParams` for them are [not even implemented](https://github.com/rust-bitcoin/bitcoin-constants/pull/1/files#diff-b4aea3e418ccdb71239b96952d9cddb6R450).

I also think the multiple traits and enums and types make it very complex. 

If we just use the Bitcoin networks, we could have three global static structs, an enum and a getter method on the enum to retrieve the struct.

sgeisler commented:
@stevenroose I agree with you that removing the partially implemented networks would simplify things since we wouldn't need so many functions, we could just have like three structs that one can get static references to. That's currently not possible because of partially missing data.

The trait on the other hand keeps the door open for possible extensions and doesn't cost anything at runtime and doesn't add too much complexity imho. And at least in `rust-lightning` modularity always seemed one of the top goals. It's possible that our priorities differ for `rust-bitcoin`.

EDIT: even the README states:
> Patches which add support for non-Bitcoin cryptocurrencies by adding constants to existing enums (e.g. to set the network message magic-byte sequence) are welcome

stevenroose commented:
Hmm. Well, I'm new to rust-bitcoin, don't want to voice my opinion too loudly. I do very much agree that it's useful to have a trait for things that are not Bitcoin-specific like B32 HRPs, xpub prefixes, ... 
Just feels like the wrong approach to go and implement all of them in the same repo. And it seems like the `from_` methods would only be useful if all coins are implemented here (or somewhere else centrally). Which would allow us to basically censor coins from using this functionality.

apoelstra commented:
A few thoughts:

1. I don't like using an `enum` of networks anywhere. There are zero situations where exhaustive checking is useful or desired because there are an indefinite number of shitcoins, many of which are missing features and/or have overlapping constants.
2. Instead I'd like each network to be a unit struct which implements the `NetworkConstants` trait, which i think can just be called `Network` because the crate already has `constants` in its name.
3. The `from_name`, `from_magic`, `from_hrp` functions can be top-level functions that return `Box<Network>`. Because the concrete structures are 0-sized this will not actually result in any allocation, though the pointer will still be fat.

I'm fine with having one massive trait. The vtable is only created once and the API becomes really complex if there are multiple traits that can't be interconverted.

Regarding which altcoins we support, I agree it's not so great that we support a tiny random collection, and it would be better to be Bitcoin-only. But I also don't feel very strongly about this and I'd rather we just retain the existing support set. Later if these projects want to implement their own -constants crate (e.g. to replace all the `unimplemented!` blocks) we can remove support from this one.

sgeisler commented:
As long as we can keep the trait in some form I'm happy. The boxing of zero sized structs is an interesting idea. 

People will still be able to match against these boxed traits but it will be obvious that there can't be exhaustive matches. And if application developers wanted to introduce an enum (e.g. to combine it with some [option parsing magic](https://github.com/TeXitoi/structopt/blob/master/examples/enum_in_args.rs)) they can still do that and e.g. implement the trait for the enum.

What I will do now:
- [x] remove the `from_*` functions from `Networks` since they make no sense without a limited set of possible variants (with a size greater than one)
- [x] implement the `Network` trait for three zero sized structs: `Bitcoin`, `BitcoinTestnet`, `BitcoinRegtest` (effectively removing litecoin and vertcoin)
- [x] implement a `from_*` as free functions (do we need all of them? or more? or other ones?)
- [x] remove the enums

Did I forget anything in this list? Do you other Ideas I should consider?

sgeisler commented on src/lib.rs:
```diff
@@ -0,0 +1,668 @@
+/// Parameters that influence chain consensus.
+pub struct ChainParams<N: NetworkConstants> {
+    /// Network for which parameters are valid.
+    pub network: N,
```
@apoelstra Why is the network needed as a member variable here. It's not too problematic but feels redundant every time I have to touch this line.

apoelstra replied:
Heh, it could be a `PhantomData<N>` I guess. I don't recall writing this code but it appears to be there to prevent different `ChainParams` from being used interchangeably. (Though, is this the right thing to do? It prevents comparisons, assignments, etc., of different network params. I can see an argument either way.)

sgeisler replied:
That might be confusing: [before I refactored the struct](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/consensus/params.rs#L47) this was just an enum, so this wasn't introduced for type safety. It seems redundant since you will typically know what network you *want* to be on and will fetch the according `ChainParams`. So, if I understand you correctly this can probably be removed?

EDIT: `Params` (the predecessor of `ChainParams`) was introduced in #93 and ever used in `rust-bitcoin` (the crate) since then.

apoelstra replied:
Ah, yeah, I think it can be removed.

sgeisler committed:
Use boxed trait objects internally instead of an enum

sgeisler commented:
I built a version that uses the boxed zero sized type idea. But it's a pain to work with trait objects and the restrictions they put on traits. For example serde somehow works with `serde_json` now but not with `serde_test` and to implement serde properly I'd need a newtype around the box or a type [erased serde trait](https://github.com/dtolnay/erased-serde) (which I really want to avoid).

I'll explore the box-in-newtype approach tomorrow.

stevenroose commented on src/lib.rs:
```diff
@@ -0,0 +1,530 @@
+
+/// Parameters that influence chain consensus.
+#[derive(Debug, Clone)]
+pub struct ChainParams {
```
Are you sure the data in this struct is useful outside of Bitcoin chains? Perhaps this can just be part of rust-bitcoin-chain and that's it?

sgeisler replied:
But then you would have the problem of matching against an unknown set of networks again and would loose all benefits gained by introducing the enum `NetworkConstants` since the set of supported networks couldn't be easily extended anymore.

stevenroose commented on src/lib.rs:
```diff
@@ -0,0 +1,530 @@
+}
+
+/// Provides network constants for a bitcoin-like crypto currency
+pub trait NetworkConstants {
```
I personally never use types before their declaration. In this case, I really think it makes sense to declare it before the module with the implementations. Just for readability sake, though.

sgeisler replied:
I agree.

stevenroose commented:
With your latest changes, what is the workflow? Does a method take a `Network` or a `&NetworkConstants`?

And what do I put in that method? Do I have to use the `::new()` and rebox every time? Is a global const Box possible?

sgeisler commented:
@stevenroose the previous version was a mess, so I won't waste time on explaining it (see below). A global const box is unfortunately not possible since we would need constant functions/compile time evaluated functions which are unstable. There is a crate called `lazy_static` that solves this problem but I don't think we want to use it unless it gains us much performance (it uses unsafe). I doubt that boxing zero sized structs is expensive since it just copies a pointer to the vtable and doesn't allocate.

 The current structure is:

* the `NetworkConstants` trait defines which constant can be queried
* the `Network` struct encapsulates a `Box<NetworkConstants>` trait object and can hold any network constants object defined either in this crate or another one
  * it has shortcuts to construct itself for bitcoin, testnet and regtest (maybe I want to add some clever trait magic to allow library authors to extend this shortcut function list)
  * it implements `Deref`to access the trait object in an ergonomic way
  * it implements `Debug`
  * it **does not** implement serde traits since there is no way to implement `Deserialize` in an extensible way without some ugly global registry for supported network structs. That's the responsibility of application authors that know which set they want to support.
* the `networks` module implements the `NetworkConstants` trait for the zero sized structs `Bitcoin`, `BitcoinTestnet` and `BitcoinRegtest`

I unfortunately have no clue what I should test since most of the stuff just passes constants around. @apoelstra what would you expect from the test suite? Maybe just calling all functions for all networks and see that they don't panic?

apoelstra commented:
Yeah, it'd be nice to call all the functions, but honestly we could probably get away with just having no tests.

sgeisler committed:
add 'no panic' tests

sgeisler committed:
Make testnet and regtest name indicate the base network

apoelstra commented on src/lib.rs:
```diff
@@ -0,0 +1,223 @@
+
+#[cfg(test)]
+mod tests {
+    use ::{Network};
```
nit: remove `{` `}`
